pallets/bitcoin_locks/src/lib.rs

```diff
@@ -958,32 +962,61 @@ pub mod pallet {
 			Ok(())
 		}
 
-		pub fn get_redemption_price(satoshis: &Satoshis) -> Result<T::Balance, Error<T>> {
-			const REDEMPTION_MULTIPLIER: FixedU128 = FixedU128::from_rational(713, 1000);
-			const REDEMPTION_ADDON: FixedU128 = FixedU128::from_rational(274, 1000);
-			let mut price: u128 = T::PriceProvider::get_bitcoin_argon_price(*satoshis)
+		pub fn get_redemption_price(
+			satoshis: &Satoshis,
+			lock_price: Option<T::Balance>,
+		) -> Result<T::Balance, Error<T>> {
+			let satoshis: u128 = (*satoshis).into();
+			let mut price: u128 = T::PriceProvider::get_latest_btc_price_in_us_cents()
 				.ok_or(Error::<T>::NoBitcoinPricesAvailable)?
-				.unique_saturated_into();
-			let cpi = T::PriceProvider::get_argon_cpi().unwrap_or_default();
+				.saturating_mul_int(satoshis)
+				// has 2 decimal places, but argons are in 6, so we need to scale it
+				.saturating_div(10_000u128);
 
-			if cpi.is_positive() {
-				let argon_price =
-					T::PriceProvider::get_latest_argon_price_in_us_cents().unwrap_or_default();
-
-				let multiplier = REDEMPTION_MULTIPLIER
-					.saturating_mul(argon_price)
-					.saturating_add(REDEMPTION_ADDON);
-
-				// Apply the formula: R = (Pb / Pa) * (0.713 * Pa + 0.274)
-				// Pa should be in a float value (1.01)
-				// `price` is already (Pb / Pa)
-				// The redemption price of the argon allocates fluctuating incentives based on how
-				// fast the dip should be incentivized to be capitalized on.
+			if let Some(lock_price) = lock_price {
+				price = price.min(lock_price.into());
+			}
 
-				price = multiplier.saturating_mul_int(price);
-			};
+			let r = T::PriceProvider::get_redemption_r_value().unwrap_or(FixedU128::one());
 
-			Ok(price.into())
+			// Case 1: If argon is at or above target price, no penalty — unlock cost is just b.
+			let price: T::Balance = if r >= FixedU128::one() {
+				price
+			}
+			// Case 2: Mild deviation (0.90 ≤ r < 1) — apply quadratic curve to scale unlock cost.
+			else if r >= FixedU128::from_rational(0_9, 1_0) {
+				const FX_20: FixedU128 = FixedU128::from_u32(20);
+				const FX_38: FixedU128 = FixedU128::from_u32(38);
+				const FX_19: FixedU128 = FixedU128::from_u32(19);
+
+				// Formula: b * (20r² - 38r + 19)
+				((FX_20 * r.saturating_pow(2)) + FX_19)
+					.ensure_sub(FX_38 * r)
+					.map_err(|_| Error::<T>::OverflowError)?
+					.saturating_mul_int(price)
+			}
+			// Case 3: Moderate deviation (0.01 ≤ r < 0.90) — apply rational linear formula.
+			else if r >= FixedU128::from_rational(0_01, 1_00) {
```

Comment:
Shouldn't this be `from_rational(0_01, 0_9)? I realize that technically this works since the previous else if catches anything above 0_9, but fixing this might improve clarity.